[Bug]: panic: unable to find signature

### Problem encountered

A new validator on the Mainnet network got a panic after restarting their node:

```log
2023-08-17T11:40:19.174+0200    PANIC   core.protocol.validators        validators/signatures.go:150    unable to find signature        {"error": "unable to find pending signature by ethereum address \"0xc4CfBAe580d452ABe563B5218cF454BeD9D5A67c\""} 
2023-08-17T11:40:19.174+0200    INFO    core.protocol   protocol/protocol.go:154        Stopping protocol services 
2023-08-17T11:40:19.174+0200    INFO    core.protocol.event-forwarder   evtforward/engine.go:153        Stopping the Ethereum Event Forwarder 
2023-08-17T11:40:19.174+0200    INFO    core.protocol.event-forwarder   evtforward/engine.go:156        The Event Forwarder engine stopped 
2023-08-17T11:40:19.174+0200    INFO    core.api.grpc   api/grpc.go:290 Gracefully stopping gRPC based API 
2023-08-17T11:40:19.174+0200    INFO    core.api.restproxy      rest/server.go:134      Stopping REST<>GRPC based API 
2023-08-17T11:40:19.174+0200    INFO    core.admin.validator-server     admin/server.go:154     Stopping Server<>RPC based API 
2023-08-17T11:40:19.723+0200    INFO    core    node/node.go:225        Vega shutdown complete  {"version": "v0.71.6", "version-hash": "7a23f5e2f0fb4981c8318253142e2e23e3aa4f7c"} 
panic: unable to find signature [recovered] 
        panic: unable to find signature 
goroutine 221 [running]: 
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1.1() 
        /home/runner/work/vega/vega/cmd/vega/commands/node/node.go:137 +0x59 
panic({0x2d3fc40, 0xc001ef21c0}) 
        /opt/hostedtoolcache/go/1.19.10/x64/src/runtime/panic.go:884 +0x212 
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x1?, 0x1?, {0x0?, 0x0?, 0xc00084c180?}) 
        /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:198 +0x65 
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000002000, {0xc005eba540, 0x1, 0x1}) 
        /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:264 +0x3ec 
go.uber.org/zap.(*Logger).Panic(0xc013834c60?, {0x35d3956?, 0x40?}, {0xc005eba540, 0x1, 0x1}) 
        /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:258 +0x59 
code.vegaprotocol.io/vega/core/validators.(*ERC20Signatures).offerValidatorAddedSignatures(0xc013834c60, {0xc00688f000?, 0xc00e0bbba0?}) 
        /home/runner/work/vega/vega/core/validators/signatures.go:150 +0x193 
code.vegaprotocol.io/vega/core/notary.(*SnapshotNotary).OfferSignatures(0x0?, 0x3, 0xc001283f90) 
        /home/runner/work/vega/vega/core/notary/snapshot.go:134 +0xbd 
code.vegaprotocol.io/vega/core/validators.(*ERC20Signatures).OfferSignatures(0xc013834c60) 
        /home/runner/work/vega/vega/core/validators/signatures.go:122 +0x6e 
code.vegaprotocol.io/vega/core/validators.(*Topology).BeginBlock(_, {_, _}, {{0xc002939be0, 0x20, 0x20}, {{0xb, 0x1}, {0xc000675398, 0x11}, ...}, ...}) 
        /home/runner/work/vega/vega/core/validators/topology.go:464 +0x370 
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc002939be0, 0x20, 0x20}, {{0xb, 0x1}, {0xc000675398, 0x11}, 0xa905ed, {0x2fd9f37, ...}, 
...}, ...}) 
        /home/runner/work/vega/vega/core/processor/abci.go:805 +0xeb5 
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc002939be0, 0x20, 0x20}, {{0xb, 0x1}, {0xc000675398, 0x11}, 0xa905ed, {0x2fd9f37, ...}, ...}, ...}) 
        /home/runner/work/vega/vega/core/blockchain/abci/abci.go:46 +0x75
```

### Observed behaviour

They `wanted to be sure we changed the ETH RPC and restarted vega visor and end up with following error`.

### Expected behaviour

the node should restart successfully.

### Steps to reproduce

```bash
It might not be easily reproducible.
```


### Software version

v0.71.6

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

```shell
the Mainnet network.
```


### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -94,6 +94,7 @@
 - [9038](https://github.com/vegaprotocol/vega/issues/9038) - New market proposals are no longer in both the active and enacted slices to prevent pointer sharing.
 - [8878](https://github.com/vegaprotocol/vega/issues/8878) - Fix amend to consider market decimals when checking for sufficient funds.
 - [8698](https://github.com/vegaprotocol/vega/issues/8698) - Final settlement rounding can be off by a 1 factor instead of just one.
+- [9113](https://github.com/vegaprotocol/vega/issues/9113) - Only add pending signatures to the retry list when the notary engine snapshot restores.
 - [8861](https://github.com/vegaprotocol/vega/issues/8861) - Fix successor proposals never leaving proposed state.
 - [9086](https://github.com/vegaprotocol/vega/issues/9086) - Set identifier generator during perpetual settlement for closed out orders.
 - [8884](https://github.com/vegaprotocol/vega/issues/8884) - Do not assume `\n` is present on the first read chunk of the input

--- a/core/notary/snapshot.go
+++ b/core/notary/snapshot.go
@@ -201,8 +201,7 @@ func (n *SnapshotNotary) restoreNotary(notary *types.Notary, p *types.Payload) e
 			n.log.Panic("invalid signature from snapshot", logging.Error(err))
 		}
 		ns := nodeSig{node: s.Node, sig: string(sig)}
-
-		if isValidator {
+		if isValidator && s.Pending {
 			signed := selfSigned[idK]
 			if !signed {
 				selfSigned[idK] = strings.EqualFold(s.Node, self)
@@ -225,6 +224,7 @@ func (n *SnapshotNotary) restoreNotary(notary *types.Notary, p *types.Payload) e
 	for resource, ok := range selfSigned {
 		if !ok {
 			// this is not signed, just add it to the retries list
+			n.log.Info("self-signature missing, adding to retry list", logging.String("id", resource.id))
 			retries.Add(resource, nil)
 		}
 	}

--- a/core/notary/snapshot.go
+++ b/core/notary/snapshot.go
@@ -141,8 +141,8 @@ func (n *SnapshotNotary) OfferSignatures(
 func (n *SnapshotNotary) serialiseNotary() ([]byte, error) {
 	sigs := make([]*types.NotarySigs, 0, len(n.sigs)) // it will likely be longer than this but we don't know yet
 	for ik, ns := range n.sigs {
+		_, pending := n.pendingSignatures[ik]
 		for _n := range ns {
-			_, pending := n.pendingSignatures[ik]
 			sigs = append(sigs,
 				&types.NotarySigs{
 					ID:      ik.id,
@@ -156,7 +156,7 @@ func (n *SnapshotNotary) serialiseNotary() ([]byte, error) {
 
 		// the case where aggregate has started but we have no node sigs
 		if len(ns) == 0 {
-			sigs = append(sigs, &types.NotarySigs{ID: ik.id, Kind: int32(ik.kind)})
+			sigs = append(sigs, &types.NotarySigs{ID: ik.id, Kind: int32(ik.kind), Pending: pending})
 		}
 	}
 

--- a/core/notary/snapshot_test.go
+++ b/core/notary/snapshot_test.go
@@ -16,6 +16,7 @@ import (
 	"bytes"
 	"context"
 	"testing"
+	"time"
 
 	"code.vegaprotocol.io/vega/core/types"
 	commandspb "code.vegaprotocol.io/vega/protos/vega/commands/v1"
@@ -144,4 +145,59 @@ func TestNotarySnapshotRoundTrip(t *testing.T) {
 	_, ok2 := notr.IsSigned(context.Background(), "resid1", types.NodeSignatureKindAssetNew)
 	require.True(t, ok1)
 	require.True(t, ok2)
+}
+
+func TestRetryPendingOnly(t *testing.T) {
+	notr := getTestNotary(t)
+	defer notr.ctrl.Finish()
+
+	notr.top.EXPECT().Len().AnyTimes().Return(1)
+	notr.top.EXPECT().IsValidatorVegaPubKey(gomock.Any()).AnyTimes().Return(true)
+	notr.top.EXPECT().IsTendermintValidator(gomock.Any()).AnyTimes().Return(true)
+	notr.top.EXPECT().IsValidator().AnyTimes().Return(true)
+	notr.top.EXPECT().SelfVegaPubKey().AnyTimes().Return("123456")
+
+	// we will start this signature aggregation but not send in a signature
+	notr.StartAggregate(
+		"resid1", types.NodeSignatureKindERC20MultiSigSignerAdded, []byte("123444"))
+
+	// we will start another one send in signatures from a pretend other validator
+	resID := "resid2"
+	notr.StartAggregate(resID, types.NodeSignatureKindAssetNew, []byte("123456"))
+	key := "other-validator"
+	ns := commandspb.NodeSignature{
+		Sig:  []byte("123456"),
+		Id:   resID,
+		Kind: types.NodeSignatureKindAssetNew,
+	}
+	err := notr.RegisterSignature(context.Background(), key, ns)
+	require.Nil(t, err)
+
+	// get the snapshot state and load into a new engine
+	state, _, err := notr.GetState(allKey)
+	require.Nil(t, err)
+
+	snap := &snapshot.Payload{}
+	err = proto.Unmarshal(state, snap)
+	require.Nil(t, err)
+
+	snapNotr := getTestNotary(t)
+	defer snapNotr.ctrl.Finish()
+	snapNotr.top.EXPECT().Len().AnyTimes().Return(1)
+	snapNotr.top.EXPECT().IsValidator().AnyTimes().Return(true)
+	snapNotr.top.EXPECT().SelfVegaPubKey().AnyTimes().Return("this-validator")
+	snapNotr.top.EXPECT().IsValidatorVegaPubKey(gomock.Any()).AnyTimes().Return(true)
+	snapNotr.top.EXPECT().IsTendermintValidator(gomock.Any()).AnyTimes().Return(true)
+
+	_, err = snapNotr.LoadState(context.Background(), types.PayloadFromProto(snap))
+	require.Nil(t, err)
+
+	s1, _, err := notr.GetState(allKey)
+	require.Nil(t, err)
+	s2, _, err := notr.GetState(allKey)
+	require.Nil(t, err)
+	require.True(t, bytes.Equal(s1, s2))
+
+	snapNotr.cmd.EXPECT().Command(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
+	snapNotr.OnTick(context.Background(), time.Now())
 }

